lib/rgit/cli.rb

```diff
@@ -64,4 +54,123 @@ def test(*args)
       puts args
     end
   end
-end
\ No newline at end of file
+end
+
+module Impl
+
+  require "pty"
+  def format_options(option_hash)
+    result = ""
+    option_hash.each { |k, v|
+      key_str = " --#{k.to_s.gsub("_", "-")}"
+      case v # v is truthy in all cases except: nil, false
+        when [true, false].include?(v) # it's a boolean
+          result << key_str
+        when String
+          result << key_str << "=#{v}"
+        when Fixnum
+          result << key_str << "=#{v}"
+        when Hash
+          v.each { |key, val|
+            result << key_str << " #{key}=#{val}"
+          }
+      end
+    }
+    result
+  end
+
+  def lowest_repo_above(start_dir)
+    res_dir = nil
+    start_dir.ascend { |dir|
+      if dir.has_child? ".git"
+        res_dir = dir
+        break
+      end
+    }
+    res_dir
+  end
+
+  def git_command(cmd, opt_str, dir)
+    run_in_shell("git #{cmd} #{opt_str}", dir)
+  end
+
+  def run_in_shell(cmd, dir)
+    `cd #{dir}; #{cmd}`
+  end
+
+  # Opens a virtual shell at the specified
+  # dir and runs the given cmd
+  def run_in_pty(cmd, dir)
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.